crates/sparrow-session/src/execution.rs

```diff
@@ -99,29 +95,51 @@ impl Execution {
 
     pub fn next_blocking(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
         self.is_done()?;
-        Ok(self.rt.block_on(self.output.next()))
+        Ok(self.handle.block_on(self.output.next()))
     }
 
-    pub async fn collect_all(self) -> error_stack::Result<Vec<RecordBatch>, Error> {
-        // TODO: For large outputs, we likely need to drain the output while waiting for the future.
-        match self.status {
-            Status::Running(future) => future.await?,
+    pub async fn collect_all(mut self) -> error_stack::Result<Vec<RecordBatch>, Error> {
+        let mut progress = match self.status {
+            Status::Running(progress) => progress,
             Status::Failed => error_stack::bail!(Error::ExecutionFailed),
-            _ => {}
+            Status::Completed => {
+                // If the progress channel has completed without error, we know that the output channel
+                // hasn't filled up, so we can go ahead and collect the output
+                return Ok(self.output.collect().await);
+            }
         };
 
-        Ok(self.output.collect().await)
+        let mut outputs = Vec::new();
+        loop {
+            tokio::select! {
+                // Poll futures in the order listed.
+                biased;
+
+                progress = progress.next() => {
+                    if let Some(execute_response) = progress {
```

Comment:
If this is `None`, then the stream is done. Do we need to do anything special in that case? Guessing no -- we just need to wait until the output stream finishes.